### ISSUE TITLE
Empty GreenPile iterator hangs

### ISSUE THREAD

**Issue by dmcnelis** [eventlet/eventlet#53]
Here is a simplified recreation of the problem:

```
from eventlet import GreenPile, GreenPool
pool = GreenPool()
pile = GreenPile(pool)
for x in pile:
    print x
```

This will just sit and spin, without throwing any sort of exception / error.

**Comment by nat-goodspeed** [comment]
I hit this one in my own code a couple days ago. It was a nasty surprise.

I reviewed all references to `GreenPile` on the master branch. Most of them are tests, and those tests unconditionally spawn at least one greenthread on the `GreenPile`.

It seems to me that the `feedscraper.py` example is susceptible to this bug if an empty (or all whitespace) body is posted. To me, that example clearly implies that iterating over a `GreenPile` with no `spawn()` calls should produce an empty sequence. Otherwise, the web application hangs.

`greenpool.GreenMap`, used by `GreenPool.starmap()` and `imap()`, is a subclass of `GreenPile`. But `starmap()` should (!) never return an "empty" `GreenMap`: it spawns `GreenPool._do_map()`, which spawns a greenthread per iterable item and then unconditionally spawns `return_stop_iteration()`. The same is true of `imap()`, which calls `starmap()`.

I think I know where the undesirable empty-`GreenPile` behavior comes from. Please sanity-check my reasoning.

[This commit from 2009](https://github.com/eventlet/eventlet/commit/3ddbba23de2ad1485d968bddbbb3e2866b4429c2), by one of the original eventlet authors, explicitly adds `GreenPile.used`. The only effect of the `used` flag is to force the bad behavior of which this bug (correctly) complains. The same commit introduces the `GreenPile` subclass (then called) `GreenImap`, reimplementing the `GreenPool` helper method (then called) `_do_imap()` to use it.

Because `spawn()` can block, `_do_imap()` is spawned as a standalone greenthread (external to the parent pool) to avoid blocking `starmap()`'s caller. But `imap()`/`starmap()` spawns `_do_imap()`, then immediately returns the `GreenImap`/`GreenMap` instance to its caller, which presumably starts trying to iterate over it. Note that this is the very scenario about which `GreenPile`'s present docstring warns: spawning in one greenthread, iterating in another. `_do_imap()` has not yet started running because nothing in `starmap()` has blocked the caller, or caused the running greenthread to yield to the hub. Without the `used` flag, `starmap()` must have consistently produced the empty sequence because `_do_imap()` had not yet even started spawning greenthreads on the `GreenMap`.

Ironically, [a subsequent commit](https://github.com/eventlet/eventlet/commit/9d07716c) by the same author a couple weeks later, in addition to renaming `GreenImap` to `GreenMap`, introduced a `next()` method override that completely disregards `GreenPile.used`.

In other words, the `GreenPile.used` flag, presumably introduced to support `GreenMap`, `GreenPool.starmap()` and `GreenPool.imap()`, causes this bug in `GreenPile` -- yet is, to this day, [ignored by `GreenMap`](https://github.com/eventlet/eventlet/blob/master/eventlet/greenpool.py#L242) and hence the `GreenPool` methods that rely on `GreenMap`.

Makes me wonder whether `GreenMap` should really be derived from `GreenPile` -- or whether, in fact, a whole new implementation of `GreenPool.starmap()` and `imap()` wouldn't be more robust.

I have to question this analysis, though, because I see no open issues about `GreenPool.starmap()` or `imap()` producing empty results with a non-empty input iterable.

**Comment by nat-goodspeed** [comment]
Ah. What saves `starmap()` and `imap()` is the fact that `GreenMap` depends on its consumer -- in this case the greenthread running `_do_map()` -- explicitly injecting `StopIteration`.

But that seems to suggest that the world would be better off without `GreenPile.used`. Will try that.
